Fix write_cdr/take_cdr with SHM

sumanth-nirmal opened the pull request:
This MR fixes the data path for `write_cdr`/`take_cdr` with SHM

### Details

1. With SHM enabled in `write_cdr`, when constructing the `serdata`, memory is loaned from `iceoryx` and copies the user-provided serialized data into this loaned memory and updates the `serdata` with the loaned chunk
2. While receiving the sample in the `CDRSamplesHolder` if the data is over iceoryx the data is taken from the iceoryx chunk, copied into the `CDRBlob` for the user, and the loaned chunk is released

Closes #175 

This should be merged after https://github.com/eclipse-cyclonedds/cyclonedds/pull/1031

sumanth-nirmal commented:
@eboasson and @MatthiasKillat Can you please take a look at the changes? I have updated with the current state of https://github.com/eclipse-cyclonedds/cyclonedds/pull/1031

sumanth-nirmal commented on src/ddscxx/include/dds/sub/detail/SamplesHolder.hpp:
```diff
@@ -139,8 +139,39 @@ class CDRSamplesHolder : public SamplesHolder
+            } else {
+              // this shouldn't never happen
+              ISOCPP_THROW_EXCEPTION(ISOCPP_PRECONDITION_NOT_MET_ERROR,
+                "The received sample over SHM is not in serialized form");
```
Probably this assumption is not always true!

sumanth-nirmal replied:
The scenario that the iox chunk containing the raw data should also be handled here, I will modify this accordingly

eboasson replied:
It's probably worth splitting this out into a separate function. Also we should do something about the `+4`s ...

MatthiasKillat replied:
Right., the 4 should be the CDR header size constant that should exist somewhere, if not has be created. 

@sumanth-nirmal I am not sure we can get raw (original, unserialized) data here, but if so yes it should be handled This case must have been handled before (fixed-size types) somewhere though ...

MatthiasKillat commented on src/ddscxx/include/dds/sub/detail/SamplesHolder.hpp:
```diff
@@ -139,8 +139,39 @@ class CDRSamplesHolder : public SamplesHolder
       {
+        struct ddsi_serdata * current_blob = cdr_blobs[i];
+#ifdef DDS_HAS_SHM
+        // if the data is over iox_chunk
```
typo? over

MatthiasKillat commented on src/ddscxx/include/dds/sub/detail/SamplesHolder.hpp:
```diff
@@ -139,8 +139,39 @@ class CDRSamplesHolder : public SamplesHolder
+            // get the size
+            auto iox_header = iceoryx_header_from_chunk(current_blob->iox_chunk);
+            // if the iox chunk has the data in serialized form
+            if (iox_header->shm_data_state == IOX_CHUNK_CONTAINS_SERIALIZED_DATA) {
```
The state can also now be obtained with `iox_shm_data_state_t shm_get_data_state(void *iox_chunk);` Recommended if the header is not required for anything else.

sumanth-nirmal replied:
I need the header for the size of the chunk, so I was not using `shm_get_data_state()`


Ideally, the interface might need to have proper setters and getters for the fields in the user header, we can probably do as a followup

MatthiasKillat commented on src/ddscxx/src/org/eclipse/cyclonedds/pub/AnyDataWriterDelegate.cpp:
```diff
@@ -22,10 +22,15 @@
+#include "dds/features.hpp"
+
+#ifdef DDSCXX_HAS_SHM
+#include <dds/ddsi/shm_transport.h>
```
Not for this PR but we probably need to expose the functionality in a C header and be able to use it without requiring the `#ifdef` here ...

MatthiasKillat commented on src/ddscxx/src/org/eclipse/cyclonedds/pub/AnyDataWriterDelegate.cpp:
```diff
@@ -116,6 +121,29 @@ AnyDataWriterDelegate::write_cdr(
+        // copy the header
+        memcpy(iox_chunk, data->encoding().data(), data->encoding().size());
+        // copy the actual data
+        memcpy(static_cast<unsigned char *>(iox_chunk) + data->encoding().size(),
```
Copying the serialized data later after it has been serialized is also a source of severe inefficiencies. Ideally it should already have been serialized to the buffer. I assume that due to the existing architecture it may be needed n other memory of the serialized sample representation as well, but I am not sure.

sumanth-nirmal replied:
Since we don't have a user API to loan the memory and fill the serialized sample, we get the serialized sample as `CDRBlob` then copy this into the `iox_chunk`, which will then be eventually transferred using shm

MatthiasKillat commented on src/ddscxx/src/org/eclipse/cyclonedds/pub/AnyDataWriterDelegate.cpp:
```diff
@@ -124,7 +152,18 @@ AnyDataWriterDelegate::write_cdr(
+    if (ret != DDS_RETCODE_OK) {
+        if (ser_data->iox_chunk != nullptr) {
+            // write cdr failed, so free the chunk
+            ret = dds_return_loan(writer, &ser_data->iox_chunk, 1);
```
We need to double check whether return loan works with `dds_loan_shared_memory_buffer`, but it should (it is just delegated to iceoryx) There are more tests needed ...

sumanth-nirmal replied:
This should be tested indirectly with the tests in the upper layers

Agree, we might need to add some tests for these new shm APIs (at some point)

sumanth-nirmal commented:
@eboasson @MatthiasKillat I think I have addressed the above issues, please take a look.

Also, I used a temporary commit (which is dropped now) to test the CI with the branch here https://github.com/eclipse-cyclonedds/cyclonedds/pull/1031 and it is green https://dev.azure.com/eclipse-cyclonedds/cyclonedds-cxx/_build/results?buildId=1527&view=results

sumanth-nirmal commented:
@eboasson I consolidated the code blocks handling the iox_chunk into API's so now it should not have a mix of pre-processing conditionals and branching, please take a look

eboasson reviewed (approved):
Thanks @sumanth-nirmal !
